The failing evaluation cases are already marked as known-to-fail in #644:
```
FAILED evals/features/cve/test_suggest_cwe.py::test_eval_suggest_cwe - AssertionError: Unsatisfied assertion(s):
suggest-cwe-for-CVE-2025-22097: SuggestCweEvaluator: score below threshold: -0.2125 < 0.1, reason: got ['CWE-364', 'CWE-772'], expected ['CWE-824', 'CWE-825']
FAILED evals/features/cve/test_suggest_impact.py::test_eval_suggest_impact - AssertionError: Unsatisfied assertion(s):
suggest-impact-for-CVE-2025-39795: CVSSKernelScopeAndPrivileges: False, reason: The explanation for Integrity Impact (I) states 'Moderate' and describes 'significant impact on data integrity', which is inconsistent with the CVSS vector's 'I:L' (Low). The rubric requires the explanation to be consistent with the vector.
```

Let's merge #644 first and rebase this onto `main`.  Please also resolve the outstanding review comments.

--- a/src/aegis_ai/features/__init__.py
+++ b/src/aegis_ai/features/__init__.py
@@ -206,6 +206,12 @@ class Feature(ABC):
                 model_settings=model_settings,
                 **retry_kwargs,
             )
+        else:
+            logger.warning(
+                "%s: output enforcement exhausted %d retries; accepting result as-is",
+                call_str,
+                _MAX_OUTPUT_ENFORCEMENT_RETRIES,
+            )
 
         # check how many input tokens were processed by the LLM
         input_tokens = result._state.usage.input_tokens

--- a/src/aegis_ai/features/cve/__init__.py
+++ b/src/aegis_ai/features/cve/__init__.py
@@ -436,7 +436,6 @@ class SuggestImpact(Feature):
         )
         if pre_clf is not None:
             deps.classifier_result = pre_clf
-            deps.kernel_tool_called = True
 
         output_schema = SuggestImpactModel.model_json_schema()
         if not is_kernel:

--- a/src/aegis_ai/features/cve/__init__.py
+++ b/src/aegis_ai/features/cve/__init__.py
@@ -416,11 +416,28 @@ class SuggestImpact(Feature):
 
             is_kernel = is_kernel_component(components)
 
+        # Eagerly run the kernel classifier so the result is available on
+        # deps for both the tool fast-path and post-processing (reconciliation,
+        # guardrails).  kernel_impact_tool will return this cached result
+        # instantly when the LLM calls it.
+        if use_kernel_classifier and is_kernel:
+            from aegis_ai.toolsets.tools.kernel_classifier import kernel_impact_classify
+
+            pre_clf = await kernel_impact_classify(
+                cve_id, static_context=resolved_static_context
+            )
+        else:
+            pre_clf = None
+
         deps = feature_deps(
             exclude_osidb_fields=["impact", "rh_cvss_score"],
             static_context=resolved_static_context if use_static else None,
             is_kernel_cve=is_kernel,
         )
+        if pre_clf is not None:
+            deps.classifier_result = pre_clf
+            deps.kernel_tool_called = True
+
         output_schema = SuggestImpactModel.model_json_schema()
         if not is_kernel:
             output_schema.get("properties", {}).pop(
@@ -443,18 +460,6 @@ class SuggestImpact(Feature):
         )
 
         run_kwargs: dict = dict(deps=deps, output_type=SuggestImpactModel)
-
-        if use_kernel_classifier and is_kernel:
-            from pydantic_ai.toolsets import FunctionToolset
-            from aegis_ai.toolsets.tools.kernel_classifier import kernel_impact_tool
-
-            kernel_tools: list = [kernel_impact_tool]
-            if get_settings().use_linux_cve_tool:
-                from aegis_ai.toolsets.tools.kernel_cves import kernel_cve_tool
-
-                kernel_tools.append(kernel_cve_tool)
-
-            run_kwargs["toolsets"] = [FunctionToolset[feature_deps](tools=kernel_tools)]
 
         result = await self.guarded_run(prompt, **run_kwargs)
         call_str = f"{self.__class__.__name__}({cve_id})"

--- a/src/aegis_ai/features/cve/kernel.py
+++ b/src/aegis_ai/features/cve/kernel.py
@@ -769,29 +769,28 @@ def apply_kpanic_cvss_override(
 
 def check_kernel_output(output, deps) -> str | None:
     """Return a retry message if kernel_impact_tool was not called for a
-    kernel CVE, or ``None`` to accept the output.
-
-    The kernel_impact_tool provides patch-level analysis (active feature
-    flags and severity class probabilities) that the LLM must incorporate
-    into its assessment.  If the tool was available but not invoked, the
-    output is rejected and the LLM is asked to retry.
-    """
+    kernel CVE, or ``None`` to accept the output."""
     use_clf = get_settings().use_kernel_classifier
     is_kernel = getattr(deps, "is_kernel_cve", False)
-    tool_called = getattr(deps, "kernel_tool_called", False)
-    clf_result = getattr(deps, "classifier_result", None)
 
-    if use_clf and is_kernel and hasattr(output, "cvss3_vector"):
-        if not tool_called:
-            logger.warning(
-                "kernel_impact_tool was not called for a kernel CVE; requesting retry"
-            )
-            return (
-                "This is a Linux kernel CVE. You MUST call kernel_impact_tool "
-                "to obtain patch-level analysis before producing your final answer."
-            )
-        if tool_called and clf_result is None:
-            logger.warning(
-                "kernel_impact_tool was called but returned no classifier data for this CVE"
-            )
-    return None
+    if not (use_clf and is_kernel and hasattr(output, "cvss3_vector")):
+        return None
+
+    clf_result = getattr(deps, "classifier_result", None)
+    if clf_result is not None:
+        return None
+
+    attempts = getattr(deps, "classifier_attempts", 0)
+    if attempts > 0:
+        logger.warning(
+            "kernel_impact_tool was called but returned no classifier data for this CVE"
+        )
+        return None
+
+    logger.warning(
+        "kernel_impact_tool was not called for a kernel CVE; requesting retry"
+    )
+    return (
+        "This is a Linux kernel CVE. You MUST call kernel_impact_tool "
+        "to obtain patch-level analysis before producing your final answer."
+    )

--- a/src/aegis_ai/features/data_models.py
+++ b/src/aegis_ai/features/data_models.py
@@ -29,10 +29,11 @@ class feature_deps:
     # enforce kernel_impact_tool usage, and by the tool itself to gate
     # non-kernel calls.
     is_kernel_cve: bool = field(default=False)
-    # Set True by exec() when it pre-computes the classifier result, or by
-    # kernel_impact_tool when the LLM calls it.  check_kernel_output reads
-    # this to distinguish "tool not called" from "tool called but no data".
-    kernel_tool_called: bool = field(default=False)
+    # Incremented by kernel_impact_tool each time the LLM invokes it.
+    # check_kernel_output reads this: 0 means the LLM never called the tool
+    # (triggers retry), >0 means it tried (accept, even if classifier_result
+    # is None).
+    classifier_attempts: int = field(default=0)
     # Pre-computed classifier result from kernel_impact_classify.  Set by
     # exec() before the LLM run; kernel_impact_tool returns it immediately
     # via its fast-path cache (mirrors the flaw_tool static_context pattern).

--- a/src/aegis_ai/features/data_models.py
+++ b/src/aegis_ai/features/data_models.py
@@ -25,16 +25,18 @@ class feature_deps:
     # Used when static_context already contains sufficient CVE data (e.g. from web API).
     static_context: Optional[Any] = field(default=None)
     # Set by SuggestImpact.exec (pre-run) or flaw_tool (at runtime) when OSIDB
-    # data indicates a kernel component.  Read by _check_output to enforce
-    # kernel_impact_tool usage.
+    # data indicates a kernel component.  Read by check_kernel_output to
+    # enforce kernel_impact_tool usage, and by the tool itself to gate
+    # non-kernel calls.
     is_kernel_cve: bool = field(default=False)
-    # Set by kernel_impact_tool when it is invoked, regardless of whether the
-    # classifier produced a result.  _check_output uses this to distinguish
-    # "tool not called" from "tool called but no data available".
+    # Set True by exec() when it pre-computes the classifier result, or by
+    # kernel_impact_tool when the LLM calls it.  check_kernel_output reads
+    # this to distinguish "tool not called" from "tool called but no data".
     kernel_tool_called: bool = field(default=False)
-    # Pre-computed classifier result (e.g. from kernel_impact_classify).
-    # Set before the agent run and read-only during execution.  Used by
-    # tools (fast-path cache) and post-processing (escalation floor).
+    # Pre-computed classifier result from kernel_impact_classify.  Set by
+    # exec() before the LLM run; kernel_impact_tool returns it immediately
+    # via its fast-path cache (mirrors the flaw_tool static_context pattern).
+    # Also read by post-processing (reconciliation, guardrails).
     classifier_result: Optional[dict] = field(default=None)
 
 

--- a/src/aegis_ai/kernel_classifier/__init__.py
+++ b/src/aegis_ai/kernel_classifier/__init__.py
@@ -18,6 +18,7 @@ Configuration (environment variables):
   AEGIS_USE_KERNEL_CLASSIFIER  — set to "true" to enable (default: false)
 """
 
+import difflib
 import importlib.util
 import json
 import logging
@@ -139,23 +140,140 @@ def _select_best_external_cvss3(cvss_scores: list[dict]) -> tuple[str, float, st
 
 _PATCH_SIZE_LIMIT = 1_000_000
 
+_BACKPORT_SIMILARITY_THRESHOLD = 0.75
 
-def _collect_patches(patches: list[tuple[str, str]]) -> list[str]:
-    """Return patch content strings, dropping any that exceed the size limit.
 
-    Large non-fix commits exist (driver imports, treewide refactors) but
-    CVE fix commits are targeted bug fixes that are well under 1 MB.
-    The guard protects against accidental inclusion of a prohibitively
-    large patch (e.g. the ~200 MB initial kernel commit).
+def _filter_oversized(
+    patches: list[tuple[str, str]],
+) -> list[tuple[str, str]]:
+    """Return patches that are within the size limit, warning about dropped ones.
+
+    Large non-fix commits exist (driver imports, treewide refactors) but CVE
+    fix commits are targeted bug fixes well under 1 MB.  The guard protects
+    against accidental inclusion of a prohibitively large patch (e.g. the
+    ~200 MB initial kernel commit).
     """
-    out: list[str] = []
+    out: list[tuple[str, str]] = []
     for commit_hash, content in patches:
         if len(content) > _PATCH_SIZE_LIMIT:
             logger.warning(
                 "Dropping oversized patch %s (%d chars)", commit_hash[:12], len(content)
             )
             continue
-        out.append(content)
+        out.append((commit_hash, content))
+    return out
+
+
+def _is_backport_of(canonical: str, candidate: str) -> bool:
+    """Return True when candidate is a stable-backport of canonical.
+
+    Uses a two-phase SequenceMatcher check: cheap quick_ratio() pre-filter
+    followed by the full ratio() only when the upper bound is promising.
+    Kernel backports of the same fix typically score 0.85-0.98; the threshold
+    of 0.75 leaves a comfortable margin below that band while excluding
+    genuinely distinct patches (which score < 0.20 in practice).
+
+    The "diff of the diff" insight: backports differ from mainline only in
+    mail headers, index object hashes, @@ hunk line-numbers, and an
+    [Upstream commit] annotation -- everything security-relevant (commit
+    message body, fix lines) is identical.
+    """
+    sm = difflib.SequenceMatcher(None, canonical, candidate)
+    return (
+        sm.quick_ratio() >= _BACKPORT_SIMILARITY_THRESHOLD
+        and sm.ratio() >= _BACKPORT_SIMILARITY_THRESHOLD
+    )
+
+
+def _patch_delta(
+    canonical_hash: str,
+    canonical_lines: list[str],
+    commit_hash: str,
+    content: str,
+) -> str:
+    """Return a compact delta string representing a backport's differences.
+
+    Uses difflib.unified_diff with n=0 (no context lines) so only the
+    changed lines appear.  Typical differences preserved:
+      - Hunk line numbers (@@ -100 vs @@ -95) -- signals code offset in
+        the target stable branch.
+      - [Upstream commit ...] annotation -- confirms cherry-pick.
+      - Additional Signed-off-by (e.g. Greg Kroah-Hartman).
+      - index object hashes (index aaa..bbb vs index ccc..ddd).
+      - Adapted fix or context lines when the backport required manual
+        adjustment (different variable names, missing prerequisite code).
+    Everything identical between patches (commit message body, the actual
+    fix +/- lines, diffstat) is elided because unified_diff only emits
+    change hunks.  For a 10 KB patch with 5 differing lines the delta is
+    ~200-500 bytes -- a ~95% reduction per duplicate.
+    """
+    backport_lines = content.splitlines(keepends=True)
+    delta_lines = list(
+        difflib.unified_diff(
+            canonical_lines,
+            backport_lines,
+            fromfile=f"canonical ({canonical_hash[:12]})",
+            tofile=f"backport ({commit_hash[:12]})",
+            n=0,
+        )
+    )
+    if delta_lines:
+        return "".join(delta_lines)
+    return f"(identical to canonical {canonical_hash[:12]})"
+
+
+def _deduplicate_patches(patches: list[tuple[str, str]]) -> list[str]:
+    """Deduplicate near-identical stable-backport patches for LLM consumption.
+
+    Kernel CVE fixes are typically one mainline commit cherry-picked into
+    multiple stable branches (6.1.y, 6.6.y, ...).  Sending all N copies to
+    the LLM wastes tokens on redundant content.  This function picks the
+    longest patch as canonical (returned in full); near-duplicates are
+    replaced by a compact delta (see ``_patch_delta``); genuinely distinct
+    patches are kept in full (see ``_is_backport_of``).  Oversized patches
+    are dropped first (see ``_filter_oversized``).
+
+    Args:
+        patches: (commit_hash, patch_content) tuples from ``_fetch_patches``.
+
+    Returns:
+        Compact list of strings suitable for LLM consumption.
+    """
+    sized = _filter_oversized(patches)
+
+    if not sized:
+        return []
+
+    if len(sized) == 1:
+        return [sized[0][1]]
+
+    out: list[str] = []
+    remaining = sized[:]
+    while remaining:
+        canonical_idx = max(range(len(remaining)), key=lambda i: len(remaining[i][1]))
+        canonical_hash, canonical_content = remaining.pop(canonical_idx)
+        canonical_lines = canonical_content.splitlines(keepends=True)
+        out.append(canonical_content)
+
+        duplicates: list[str] = []
+        unmatched: list[tuple[str, str]] = []
+        for commit_hash, content in remaining:
+            if _is_backport_of(canonical_content, content):
+                duplicates.append(
+                    _patch_delta(canonical_hash, canonical_lines, commit_hash, content)
+                )
+            else:
+                unmatched.append((commit_hash, content))
+
+        if duplicates:
+            header = (
+                f"[{len(duplicates)} additional backport commit(s) with "
+                f"near-identical diffs — only differences shown]\n"
+            )
+            out.append(header + "\n---\n".join(duplicates))
+
+        remaining = unmatched
+
     return out
 
 
@@ -518,7 +636,7 @@ class KernelImpactClassifier:
             "cvss_score": cvss_score,
             "cvss_issuer": cvss_issuer,
             "patches_analyzed": len(patches),
-            "patch_summaries": _collect_patches(patches),
+            "patch_summaries": _deduplicate_patches(patches),
         }
         if html_supplemented_flags:
             result["html_supplemented_flags"] = html_supplemented_flags

--- a/src/aegis_ai/toolsets/__init__.py
+++ b/src/aegis_ai/toolsets/__init__.py
@@ -164,9 +164,19 @@ public_toolset = CombinedToolset(public_toolset_list)
 
 
 # Toolset containing rh specific tooling for CVE
-redhat_cve_toolset_list = [
+redhat_cve_toolset_list: list[AbstractToolset[Any]] = [
     osidb_toolset,
 ]
+
+if get_settings().use_kernel_classifier:
+    from aegis_ai.toolsets.tools.kernel_classifier import kernel_impact_tool
+
+    redhat_cve_toolset_list.append(FunctionToolset(tools=[kernel_impact_tool]))
+
+if get_settings().use_linux_cve_tool:
+    from aegis_ai.toolsets.tools.kernel_cves import kernel_cve_tool
+
+    redhat_cve_toolset_list.append(FunctionToolset(tools=[kernel_cve_tool]))
 
 redhat_cve_toolset = CombinedToolset(redhat_cve_toolset_list)
 

--- a/src/aegis_ai/toolsets/tools/kernel_classifier/__init__.py
+++ b/src/aegis_ai/toolsets/tools/kernel_classifier/__init__.py
@@ -168,7 +168,7 @@ async def kernel_impact_tool(
     signals: which patch feature flags fired and the model's per-class severity
     probabilities.  Use this when the CVE component is the Linux kernel."""
 
-    ctx.deps.kernel_tool_called = True
+    ctx.deps.classifier_attempts += 1
 
     if not ctx.deps.is_kernel_cve:
         return KernelImpactToolResponse(

--- a/src/aegis_ai/toolsets/tools/kernel_classifier/__init__.py
+++ b/src/aegis_ai/toolsets/tools/kernel_classifier/__init__.py
@@ -168,8 +168,24 @@ async def kernel_impact_tool(
     signals: which patch feature flags fired and the model's per-class severity
     probabilities.  Use this when the CVE component is the Linux kernel."""
 
-    logger.info("Analysing kernel patch features for %s...", input.cve_id)
     ctx.deps.kernel_tool_called = True
+
+    if not ctx.deps.is_kernel_cve:
+        return KernelImpactToolResponse(
+            cve_id=input.cve_id,
+            status="error",
+            error_message="Not a kernel CVE; tool not applicable.",
+        )
+
+    # Fast path: return pre-computed result when available (mirrors flaw_tool
+    # static_context pattern).  exec() eagerly runs kernel_impact_classify
+    # before the LLM call and stores the result on deps.
+    if ctx.deps.classifier_result is not None:
+        logger.info("Using pre-computed classifier result for %s", input.cve_id)
+        return _response_from_result(input.cve_id, ctx.deps.classifier_result)
+
+    # Slow path: run classifier on demand
+    logger.info("Analysing kernel patch features for %s...", input.cve_id)
     static_context = getattr(ctx.deps, "static_context", None)
     result = await kernel_impact_classify(input.cve_id, static_context=static_context)
 
@@ -180,8 +196,5 @@ async def kernel_impact_tool(
             error_message="Kernel classifier could not produce a result for this CVE.",
         )
 
-    # Store the full result (including impact) on deps so post-processing
-    # (escalation floor) can read it after the agent run completes.
     ctx.deps.classifier_result = result
-
     return _response_from_result(input.cve_id, result)

--- a/tests/test_deduplicate_patches.py
+++ b/tests/test_deduplicate_patches.py
@@ -1,0 +1,338 @@
+"""Tests for _deduplicate_patches in aegis_ai.kernel_classifier."""
+
+from aegis_ai.kernel_classifier import (
+    _PATCH_SIZE_LIMIT,
+    _deduplicate_patches,
+    _is_backport_of,
+)
+
+# ---------------------------------------------------------------------------
+# Synthetic patch fixtures
+# ---------------------------------------------------------------------------
+
+_CANONICAL = """\
+From 9e1c8c2a33d0 Mon Sep 17 00:00:00 2001
+From: Alice <alice@kernel.org>
+Date: Mon, 1 Jan 2026 12:00:00 +0000
+Subject: [PATCH] net: fix UAF in foo_handler()
+
+A use-after-free occurs in foo_handler() because kfree() is called
+before use().  The object lifetime is managed by the caller, so the
+handler must not free it.  Reorder the statements so that use()
+precedes kfree(), ensuring the object remains valid for the duration
+of the handler.
+
+The bug was introduced in commit aabbccdd1234 ("net: refactor
+foo_handler lifecycle") and is reproducible on all supported kernel
+versions when the device is hot-unplugged under load.
+
+Fixes: aabbccdd1234 ("net: refactor foo_handler lifecycle")
+Cc: stable@vger.kernel.org
+Reported-by: Eve <eve@example.org>
+Tested-by: Mallory <mallory@example.org>
+Reviewed-by: Trent <trent@example.org>
+Signed-off-by: Alice <alice@kernel.org>
+---
+ net/core/foo.c | 18 +++++++++---------
+ 1 file changed, 9 insertions(+), 9 deletions(-)
+
+diff --git a/net/core/foo.c b/net/core/foo.c
+index aaaa..bbbb 100644
+--- a/net/core/foo.c
++++ b/net/core/foo.c
+@@ -100,18 +100,18 @@ static void foo_handler(struct device *dev)
+ \tstruct foo *p = dev->priv;
+ \tstruct bar *b = p->bar;
+ \tint ret;
+ 
+-\tkfree(p);
+-\tuse(p);
++\tuse(p);
++\tkfree(p);
+ 
+ \tret = validate(b);
+ \tif (ret < 0) {
+ \t\tpr_err("validation failed: %d\\n", ret);
+ \t\treturn;
+ \t}
+ 
+ \tnotify_peer(b);
+ \tschedule_cleanup(dev);
+ 
+ \treturn;
+ }
+"""
+
+_BACKPORT_SIMILAR = """\
+From 2f03dafea0a8 Mon Sep 17 00:00:00 2001
+From: Alice <alice@kernel.org>
+Date: Mon, 1 Jan 2026 12:00:00 +0000
+Subject: [PATCH] net: fix UAF in foo_handler()
+
+[ Upstream commit 9e1c8c2a33d0 ]
+
+A use-after-free occurs in foo_handler() because kfree() is called
+before use().  The object lifetime is managed by the caller, so the
+handler must not free it.  Reorder the statements so that use()
+precedes kfree(), ensuring the object remains valid for the duration
+of the handler.
+
+The bug was introduced in commit aabbccdd1234 ("net: refactor
+foo_handler lifecycle") and is reproducible on all supported kernel
+versions when the device is hot-unplugged under load.
+
+Fixes: aabbccdd1234 ("net: refactor foo_handler lifecycle")
+Cc: stable@vger.kernel.org
+Reported-by: Eve <eve@example.org>
+Tested-by: Mallory <mallory@example.org>
+Reviewed-by: Trent <trent@example.org>
+Signed-off-by: Alice <alice@kernel.org>
+Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
+---
+ net/core/foo.c | 18 +++++++++---------
+ 1 file changed, 9 insertions(+), 9 deletions(-)
+
+diff --git a/net/core/foo.c b/net/core/foo.c
+index cccc..dddd 100644
+--- a/net/core/foo.c
++++ b/net/core/foo.c
+@@ -95,18 +95,18 @@ static void foo_handler(struct device *dev)
+ \tstruct foo *p = dev->priv;
+ \tstruct bar *b = p->bar;
+ \tint ret;
+ 
+-\tkfree(p);
+-\tuse(p);
++\tuse(p);
++\tkfree(p);
+ 
+ \tret = validate(b);
+ \tif (ret < 0) {
+ \t\tpr_err("validation failed: %d\\n", ret);
+ \t\treturn;
+ \t}
+ 
+ \tnotify_peer(b);
+ \tschedule_cleanup(dev);
+ 
+ \treturn;
+ }
+"""
+
+_DISTINCT_PATCH = """\
+From ffffffffffffffff Mon Sep 17 00:00:00 2001
+From: Bob <bob@kernel.org>
+Date: Tue, 2 Jan 2026 08:00:00 +0000
+Subject: [PATCH] sched: fix deadlock in task_migrate()
+
+Completely different fix for a scheduling deadlock.
+
+Signed-off-by: Bob <bob@kernel.org>
+---
+ kernel/sched/core.c | 10 ++++------
+ 1 file changed, 4 insertions(+), 6 deletions(-)
+
+diff --git a/kernel/sched/core.c b/kernel/sched/core.c
+index 1111..2222 100644
+--- a/kernel/sched/core.c
++++ b/kernel/sched/core.c
+@@ -500,10 +500,8 @@ static int task_migrate(struct task_struct *p)
+ \traw_spin_lock(&rq->lock);
+-\tif (p->state == TASK_DEAD) {
+-\t\traw_spin_unlock(&rq->lock);
+-\t\treturn -ESRCH;
+-\t}
++\tif (p->state == TASK_DEAD)
++\t\tgoto out_unlock;
+ \tp->on_rq = 0;
++out_unlock:
+ \traw_spin_unlock(&rq->lock);
+ \treturn 0;
+ }
+"""
+
+_SECOND_CANONICAL = """\
+From 7ab4c9d1e2f3 Mon Sep 17 00:00:00 2001
+From: Carol <carol@kernel.org>
+Date: Wed, 3 Jan 2026 09:30:00 +0000
+Subject: [PATCH] mm: fix refcount leak in baz_release()
+
+A reference leak happens when baz_release() returns early after a failed
+cleanup path. Ensure the final put_ref() runs on the error path too so
+the object lifetime is balanced across all exits.
+
+Fixes: 112233445566 ("mm: simplify baz_release cleanup")
+Cc: stable@vger.kernel.org
+Signed-off-by: Carol <carol@kernel.org>
+---
+ mm/baz.c | 6 +++---
+ 1 file changed, 3 insertions(+), 3 deletions(-)
+
+diff --git a/mm/baz.c b/mm/baz.c
+index 5555..6666 100644
+--- a/mm/baz.c
++++ b/mm/baz.c
+@@ -44,10 +44,10 @@ static int baz_release(struct baz *baz)
+-\tif (cleanup_failed(baz))
+-\t\treturn -EINVAL;
++\tif (cleanup_failed(baz))
++\t\tgoto out_put;
+ \tflush_work(&baz->work);
+-\tput_ref(baz);
+-\treturn 0;
++out_put:
++\tput_ref(baz);
++\treturn cleanup_failed(baz) ? -EINVAL : 0;
+ }
+"""
+
+_SECOND_BACKPORT_SIMILAR = """\
+From 8bc5d0e2f304 Mon Sep 17 00:00:00 2001
+From: Carol <carol@kernel.org>
+Date: Wed, 3 Jan 2026 09:30:00 +0000
+Subject: [PATCH] mm: fix refcount leak in baz_release()
+
+[ Upstream commit 7ab4c9d1e2f3 ]
+
+A reference leak happens when baz_release() returns early after a failed
+cleanup path. Ensure the final put_ref() runs on the error path too so
+the object lifetime is balanced across all exits.
+
+Fixes: 112233445566 ("mm: simplify baz_release cleanup")
+Cc: stable@vger.kernel.org
+Signed-off-by: Carol <carol@kernel.org>
+Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
+---
+ mm/baz.c | 6 +++---
+ 1 file changed, 3 insertions(+), 3 deletions(-)
+
+diff --git a/mm/baz.c b/mm/baz.c
+index 7777..8888 100644
+--- a/mm/baz.c
++++ b/mm/baz.c
+@@ -41,10 +41,10 @@ static int baz_release(struct baz *baz)
+-\tif (cleanup_failed(baz))
+-\t\treturn -EINVAL;
++\tif (cleanup_failed(baz))
++\t\tgoto out_put;
+ \tflush_work(&baz->work);
+-\tput_ref(baz);
+-\treturn 0;
++out_put:
++\tput_ref(baz);
++\treturn cleanup_failed(baz) ? -EINVAL : 0;
+ }
+"""
+
+
+class TestDeduplicatePatches:
+    def test_empty_input(self):
+        assert _deduplicate_patches([]) == []
+
+    def test_single_patch(self):
+        result = _deduplicate_patches([("abc123", _CANONICAL)])
+        assert len(result) == 1
+        assert result[0] == _CANONICAL
+
+    def test_two_identical_patches(self):
+        patches = [
+            ("abc123", _CANONICAL),
+            ("def456", _CANONICAL),
+        ]
+        result = _deduplicate_patches(patches)
+        assert len(result) == 2
+        assert result[0] == _CANONICAL
+        assert "identical" in result[1].lower() or "backport" in result[1].lower()
+
+    def test_near_duplicate_produces_delta(self):
+        patches = [
+            ("9e1c8c2a33d0", _CANONICAL),
+            ("2f03dafea0a8", _BACKPORT_SIMILAR),
+        ]
+        result = _deduplicate_patches(patches)
+        assert len(result) == 2
+        # The longest patch is selected as canonical (returned in full).
+        longest = max(_CANONICAL, _BACKPORT_SIMILAR, key=len)
+        assert result[0] == longest
+        # The delta entry references the shorter patch's commit hash and
+        # is smaller than the full raw content of the deduplicated patch.
+        delta = result[1]
+        assert "backport" in delta.lower() or "additional" in delta.lower()
+        shorter_hash = "9e1c8c2a" if longest == _BACKPORT_SIMILAR else "2f03dafe"
+        assert shorter_hash in delta
+        shorter_raw = _CANONICAL if longest == _BACKPORT_SIMILAR else _BACKPORT_SIMILAR
+        assert len(delta) < len(shorter_raw)
+
+    def test_distinct_patch_kept_in_full(self):
+        patches = [
+            ("9e1c8c2a33d0", _CANONICAL),
+            ("ffffffff", _DISTINCT_PATCH),
+        ]
+        result = _deduplicate_patches(patches)
+        # Both patches should appear in full (below similarity threshold).
+        full_text = "\n".join(result)
+        assert "foo_handler" in full_text
+        assert "task_migrate" in full_text
+
+    def test_mixed_similar_and_distinct(self):
+        patches = [
+            ("9e1c8c2a33d0", _CANONICAL),
+            ("2f03dafea0a8", _BACKPORT_SIMILAR),
+            ("ffffffff", _DISTINCT_PATCH),
+        ]
+        result = _deduplicate_patches(patches)
+        full_text = "\n".join(result)
+        # Both UAF fix and scheduling fix content should be present.
+        assert "foo_handler" in full_text
+        assert "task_migrate" in full_text
+        # One of the two similar patches is deduped (delta only), so total
+        # output must be smaller than all three raw patches concatenated.
+        raw_total = len(_CANONICAL) + len(_BACKPORT_SIMILAR) + len(_DISTINCT_PATCH)
+        assert len(full_text) < raw_total
+
+    def test_oversized_patch_dropped(self):
+        huge = "x" * (_PATCH_SIZE_LIMIT + 1)
+        patches = [
+            ("abc123", _CANONICAL),
+            ("huge000", huge),
+        ]
+        result = _deduplicate_patches(patches)
+        assert len(result) == 1
+        assert result[0] == _CANONICAL
+
+    def test_canonical_is_longest(self):
+        short = _CANONICAL[:200]
+        patches = [
+            ("short00", short),
+            ("9e1c8c2a33d0", _CANONICAL),
+        ]
+        result = _deduplicate_patches(patches)
+        # The longer patch (CANONICAL) should be the first entry.
+        assert result[0] == _CANONICAL
+
+    def test_is_backport_of_threshold(self):
+        """_is_backport_of correctly identifies backports vs distinct patches."""
+        assert _is_backport_of(_CANONICAL, _BACKPORT_SIMILAR) is True
+        assert _is_backport_of(_CANONICAL, _DISTINCT_PATCH) is False
+        assert _is_backport_of(_CANONICAL, _CANONICAL) is True
+
+    def test_multiple_backport_families_are_clustered_independently(self):
+        patches = [
+            ("7ab4c9d1e2f3", _SECOND_CANONICAL),
+            ("8bc5d0e2f304", _SECOND_BACKPORT_SIMILAR),
+            ("9e1c8c2a33d0", _CANONICAL),
+            ("2f03dafea0a8", _BACKPORT_SIMILAR),
+        ]
+        result = _deduplicate_patches(patches)
+        full_text = "\n".join(result)
+
+        # Each independent fix family should keep one full canonical patch and
+        # compact its stable backport into a delta block.
+        assert any(patch in result for patch in (_CANONICAL, _BACKPORT_SIMILAR))
+        assert any(
+            patch in result for patch in (_SECOND_CANONICAL, _SECOND_BACKPORT_SIMILAR)
+        )
+        assert full_text.count("additional backport commit(s)") == 2
+        assert "9e1c8c2a33d0" in full_text
+        assert "7ab4c9d1e2f" in full_text
+        assert len(full_text) < sum(len(content) for _, content in patches)

